File: 21560 - Sai Suraj/week 6/208_Implement Trie (Prefix Tree).py
class Trie:

    def __init__(self):
        self.trie = {}

    def insert(self, word: str) -> None:
        contents = self.trie
        word+="#"
        for c in word:
            if c not in contents:
                contents[c] ={}
            contents=contents[c]

    def search(self, word: str) -> bool:
        contents = self.trie
        word+="#"
        for c in word:
            if c not in contents:
                return False
            contents = contents[c]
        return True

    def startsWith(self, prefix: str) -> bool:
        contents= self.trie
        for c in prefix:
            if c not in contents:
                return False
            contents = contents[c]
        return True


# Your Trie object will be instantiated and called as such:
# obj = Trie()
# obj.insert(word)
# param_2 = obj.search(word)
# param_3 = obj.startsWith(prefix)

File: 21561 - Suvam Tamang/Week 6/208_Implement Trie (Prefix Tree).java
class Trie {

    TrieNode head = new TrieNode();
    
    class TrieNode {
        HashMap<Character, TrieNode> children = new HashMap<>();
        boolean isEnd = false;
    }
    
    public void insert(String word) {
        if(word == null)
            return;
        TrieNode node = head;
        for(Character ch : word.toCharArray()) {
            if(!node.children.containsKey(ch))
                node.children.put(ch, new TrieNode());
            node = node.children.get(ch);
        }
        node.isEnd = true;
    }
    
    public boolean search(String word) {
        if(word == null)
            return false;
        TrieNode node = head;
        for(Character ch : word.toCharArray()) {
            if(!node.children.containsKey(ch))
                return false;
            node = node.children.get(ch);
        }
        return node.isEnd;
    }
    
    public boolean startsWith(String prefix) {
        if(prefix == null)
            return false;
        TrieNode node = head;
        for(char ch : prefix.toCharArray()) {
            if(!node.children.containsKey(ch)){
                return false;
            } 
            node = node.children.get(ch);
            
        }
        return true; 
    }
    
}

/**
 * Your Trie object will be instantiated and called as such:
 * Trie obj = new Trie();
 * obj.insert(word);
 * boolean param_2 = obj.search(word);
 * boolean param_3 = obj.startsWith(prefix);
 */
